PythonEndProject.py

```diff
@@ -0,0 +1,153 @@
+    csv_path = os.path.join(path, output_file)
+    data_frame = pd.DataFrame(data_list)
+    data_frame.to_csv(csv_path, index=False)         
+
+def exitConfirmation():
+    while True:
+        answer = input("Are you sure? (y/n)")
+        if answer == "n":
+            return False
+        if answer == "y":
+            return True
+
+def getAgesSum(people_data, current_sum):
+    current_sum += int(people_data[-1].getAge())
+    return current_sum
+
+def notNumError(user_input, key):
+    print("Error: " + key + " must be a number. " + str(user_input) + " is not a number")
+
+def saveNewEntry(info_dicsh, info_list): # 1
+    id = input("ID: ")
+    if id in info_dicsh:
+        print("Error: ID already exists: " + str(info_dicsh[id]))
+        return False 
+    if not id.isdigit():
+        notNumError(id, "ID")
+        return False
+    name = input("Name: ")
+    age = input("Age: ")
+    if not age.isdigit():
+        notNumError(age, "Age")
+        return False
+    while True:
+        employ_or_student = input("Is " + name + " an enployee or a student (for neither type 'n'): ")
+        if employ_or_student == "n":
+            new_person = Person(id, name, age)
+            break
+        elif employ_or_student == "employee":
+            new_person = Employee(id, name, age)
+            break
+        elif employ_or_student == "student":
+            new_person = Student(id, name, age)
+            break
+    info_list.append(new_person)
+    info_dicsh[id] = new_person
+    print("ID [" + str(id) + "] saved successfuly")
+    return True
+
+def searchById(info_dicsh): #3
+    id_to_search = input("Please enter the ID you want to look for: ")
+    try:
+        info_dicsh[id_to_search].printPersonData()
+    except:
+        print("An Error has occurred please try again")
+
+def printAllEntrys(info_list):
+    for idx, entry in enumerate(info_list):
+        print(str(idx) + ". ")
+        entry.printPersonData()
+
+def printAgesAvg(sum, people_amount):
+    ages_avg = sum / people_amount
+    print(ages_avg)
+
+def printAllNames(info_list):
+    for idx, name in enumerate(info_list):
+        print(str(idx) + ". " + name.getName())
+
+def printAllIds(info_list):
+    for idx, id in enumerate(info_list):
+        print(str(idx) + ". " + id.getId())
+
+def printEntryByIndex(info_list):
+    idx = input("Please enter the index of the entry you want to print: ")
+    try:
+        idx = int(idx)
+    except ValueError:
+        notNumError(idx, "Index")
+        return
+    try:
+        info_list[idx].printPersonData()
+    except IndexError:
+        print("IndexError: list index out of range. the maximum index allowed is " + str(len(info_list) - 1))
+
+id_list = []
+id_dicsh ={}
+ages_sum = 0
+main_path = "C:\\Users\\treec\\Desktop\\forward\\Python\\projects\\advanced\\end project"
+
+while True:
+    try:
+        menu()
+    except:
+        print("Error: Option [" + str(choice) + "] does not exist. Please try again")

```

Comment:
האקספט הזה נחמד, אבל לא ברור אם הוא קופץ כי
1. האפשרות היא לא מספר
2. האפשרות פשוט לא בטווח

בנוסף, האקספט הזה לא עושה continue או מדלג בדרך כלשהי על האיפים - ולכן אף על פי שקפצה שגיאה אתה עדיין תרוץ על כל השורות שמתחת לאקספשן. זו לא ההתנהגות שאנחנו רוצים. זה בסדר לעשות כאן continue.

איך מטפלים בכמה סוגים של אקספשנים?

except סוג אקספשן ראשון:
----- handling it
מיד לאחר מכן כותבים את האקספשן השני שאתה רוצה לטפל בו
except סוג אקספשן שני:
----- handling it
ואם אתה רוצה שלא משנה מה קורה, אם נפלת לאחד מהם - אתה רוצה לעשות אותו דבר (למשל קונטיניו) אז אפשר לכתוב
finally:
----- continue

לדוגמה

except TypeError:
----- print("Error with type")
except ValueError:
----- print("Error with value")
finally:
----- continue